### ISSUE TITLE
Time measured is less than inference time given by validation

### ISSUE THREAD

**Issue by jojo9-cmd** [ultralytics/ultralytics#23950]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hi there! I have a problem understanding what is the difference between measuring the prediction method and compare it to the inference time measured by validation or benchmark. Validation and benchmark inference are nearly the same. This is my code: 

import  time

import  torch

from  ultralytics  import  YOLO

from  ultralytics.utils.benchmarksimport  benchmark

from  pathlib  import  Path

import  numpy asnp

import  csv

import  cv2

import  matplotlib

def  benchmark_time(model_name, device, logfile="benchmark_log2.csv"):

# -------- settings --------

image_folders=
["/home/images"]

images_per_run=
 128

num_runs=
 7

print("CUDA
verfügbar:",
torch.cuda.is_available())

model=
 YOLO(model_name)

image_paths=
[]

forfolder  in
image_folders:

image_paths.extend(list(Path(folder).glob("*")))

images=
[cv2.imread(str(p))
for  p
inimage_paths]

# -------- Warmup --------

print("Warmup...")

model.predict(source=images[:100],
device=device, verbose=False)

# -------- Benchmark --------

times=
[]

print("Benchmarking...")

fori
 in range(num_runs):

images_run=
 images[i*
 images_per_run:(i+1)
*  images_per_run]

t0=
 time.perf_counter()

model.predict(source=images_run, device=device, batch=1,
imgsz=640,
verbose=False)

ifdevice  !=
"cpu":

torch.cuda.synchronize()

diff=
 time.perf_counter() -
 t0

times.append(diff)

print(f"run {i+1}:
{diff:.2f} s")

time_per_image_ms=
(diff/images_per_run)*1000

withopen(logfile, "a",
newline="")
as  f:

writer=
 csv.writer(f)

writer.writerow([i+1,
diff, time_per_image_ms, device, model_name])

# -------- results ---------

times=
 np.array(times)

avg_time=
 times.mean()

median_time=
 np.median(times)

print("\n======================")

print(f"Model: {model_name}")

print(f"Device: {device}")

print(f"Runs: {num_runs}")

print(f"Images per
run: {images_per_run}")

print(f"Mean time
per run: {avg_time:.2f} s")

print(f"median of
all runs {median_time:.2f} s")

def  benchmark_acc(model_names, device, data_yaml="coco128.yaml", logfile="benchmark_acc2.csv", plot=True):

# ------- logging file --------

ifnot
 Path(logfile).exists():

withopen(logfile, "w",
newline="")
as  f:

writer=
 csv.writer(f)

writer.writerow(["mAP50-95", "mAP_class", "precision", "recall", "class_id", "model", "device"])

form
 in model_names:

# ------- model -------

model=
 YOLO(m)

# ------- evaluate model --------

metrics=
 model.val(data=data_yaml, device=device, batch=1,
half=False, verbose=False)

mAP50_95=
 metrics.box.map

class_mAP=
 metrics.box.maps

precision=
 metrics.box.p

recall=
 metrics.box.r

withopen(logfile, "a",
newline="")
as  f:

writer=
 csv.writer(f)

forclass_id, mAP
 in enumerate(class_mAP):

writer.writerow([mAP50_95, mAP,
class_id, m,
device])

if  __name__  ==
"__main__":

benchmark_time("yolo26n.pt", 0)

benchmark_time("yolo11n.pt", 0)

benchmark_time("yolov5n.pt", 0)

benchmark(model="yolo26n.pt", data="coco128.yaml", imgsz=640,
half=False, device=0)

benchmark(model="yolo11n.pt", data="coco128.yaml", imgsz=640,
half=False, device=0)

benchmark(model="yolov5n.pt", data="coco128.yaml", imgsz=640,
half=False, device=0)

benchmark_acc(model_names=["yolo26n.pt", "yolo11n.pt", "yolov5n.pt"], device=0)

Results:
Warmup...
Benchmarking...
Durchlauf 1: 0.27 s
Durchlauf 2: 0.26 s
Durchlauf 3: 0.26 s
Durchlauf 4: 0.26 s
Durchlauf 5: 0.26 s
Durchlauf 6: 0.26 s
Durchlauf 7: 0.26 s

======================
Model: yolo26n.pt
Device: 0
Runs: 7
Images per run: 128
Mean time per run: 0.26 s
median of all runs 0.26 s

Warmup...
Benchmarking...
Durchlauf 1: 0.28 s
Durchlauf 2: 0.28 s
Durchlauf 3: 0.28 s
Durchlauf 4: 0.28 s
Durchlauf 5: 0.28 s
Durchlauf 6: 0.28 s
Durchlauf 7: 0.28 s

======================
Model: yolo11n.pt
Device: 0
Runs: 7
Images per run: 128
Mean time per run: 0.28 s
median of all runs 0.28 s

PRO TIP 💡 Replace 'model=yolov5n.pt' with new 'model=yolov5nu.pt'.
YOLOv5 'u' models are trained with https://github.com/ultralytics/ultralytics and feature improved performance vs standard YOLOv5 models trained with https://github.com/ultralytics/yolov5.

Warmup...
Benchmarking...
Durchlauf 1: 0.27 s
Durchlauf 2: 0.27 s
Durchlauf 3: 0.27 s
Durchlauf 4: 0.27 s
Durchlauf 5: 0.27 s
Durchlauf 6: 0.27 s
Durchlauf 7: 0.27 s

======================
Model: yolov5n.pt
Device: 0
Runs: 7
Images per run: 128
Mean time per run: 0.27 s
median of all runs 0.27 s
Ultralytics 8.4.19 🚀 Python-3.12.3 torch-2.12.0.dev20260302+cu128 CUDA:0 (NVIDIA GeForce RTX 5080, 15834MiB)
YOLO26n summary (fused): 122 layers, 2,408,932 parameters, 0 gradients, 5.4 GFLOPs
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 5632.4±1487.1 MB/s, size: 83.6 KB)
val: Scanning /home/Documents/datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 128/128 48.8Mit/s 0.0s
                 Class     Images  Instances      Box(P        R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 128/128 126.9it/s 1.0s
                   all        128        929      0.668      0.554    0.652       0.48
Speed: 0.3ms preprocess, 5.9ms inference, 0.0ms loss, 0.1ms postprocess per image
Results saved to /home/Documents/runs/detect/val22
Ultralytics 8.4.19 🚀 Python-3.12.3 torch-2.12.0.dev20260302+cu128 CUDA:0 (NVIDIA GeForce RTX 5080, 15834MiB)
YOLO11n summary (fused): 100 layers, 2,616,248 parameters, 0 gradients, 6.5 GFLOPs
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 4177.5±1605.6 MB/s, size: 36.8 KB)
val: Scanning /home/Documents/datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 128/128 89.5Mit/s 0.0s
                 Class     Images  Instances      Box(P        R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 128/128 182.8it/s 0.7s
                   all        128        929      0.667      0.599    0.675       0.51
Speed: 0.2ms preprocess, 3.2ms inference, 0.0ms loss, 0.4ms postprocess per image
Results saved to /home/Documents/runs/detect/val23
PRO TIP 💡 Replace 'model=yolov5n.pt' with new 'model=yolov5nu.pt'.
YOLOv5 'u' models are trained with https://github.com/ultralytics/ultralytics and feature improved performance vs standard YOLOv5 models trained with https://github.com/ultralytics/yolov5.

Ultralytics 8.4.19 🚀 Python-3.12.3 torch-2.12.0.dev20260302+cu128 CUDA:0 (NVIDIA GeForce RTX 5080, 15834MiB)
YOLOv5n summary (fused): 84 layers, 2,649,200 parameters, 0 gradients, 7.7 GFLOPs
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 4208.7±2011.0 MB/s, size: 40.9 KB)
val: Scanning /home/Documents/datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 128/128 107.4Mit/s 0.0s
                 Class     Images  Instances      Box(P        R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 128/128 193.9it/s 0.7s
                   all        128        929      0.596      0.522    0.575      0.419
Speed: 0.2ms preprocess, 3.1ms inference, 0.0ms loss, 0.4ms postprocess per image

**Comment by UltralyticsAssistant** [comment]
👋 Hello @jojo9-cmd, thank you for your interest in Ultralytics 🚀! This is an automated response to help get your issue moving, and an Ultralytics engineer will also assist you soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

Since this appears to be a timing/benchmarking ❓ question, thanks for already sharing code and outputs 🙌. To help the team investigate faster, please also confirm:
- Whether your measured `predict()` timing is intended to include preprocessing, batching, data transfer, and postprocessing
- The exact command or script section you want compared against `val()` or `benchmark()`
- Whether the same input source, batch size, image size, precision, and device settings are being used across all tests
- If possible, a simplified standalone script that reproduces the comparison in the smallest form

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
Your test code is running batched inference, but validation is running single image inference (`batch=1`). Setting `batch=1` during inference doesn't have an effect because you're using list of numpy arrays. It will always run on the whole batch. `batch` argument is for when running inference on video or folder.

**Comment by goelsameer** [comment]
The difference comes from what exactly is being measured and how batching is handled.

* `predict()` measures end-to-end latency (preprocess + inference + postprocess + Python overhead)
* `val()` reports these stages separately, and its inference time isolates the forward pass more cleanly

In your specific case, the key detail is that you're passing a list of NumPy images to `predict()`. In Ultralytics YOLO, this is treated as a batch internally, so `batch=1` is effectively ignored and the model processes the images in a batched/optimized way.

On the other hand, `val()` uses its own dataloader and batching logic, so you're not comparing the same execution pattern.

That’s why your `predict()` timing (~2 ms/image) appears faster than the validation inference time (~3–6 ms/image): you're effectively measuring batched throughput vs more controlled per-stage timing.

To make them comparable, you can either:

* run `predict()` on single images in a loop (true batch=1), or
* increase batch size in `val()` to match throughput conditions

**Comment by glenn-jocher** [comment]
Agreed — one extra detail is that `benchmark()` is expected to be close to `val()` here because it calls validation with `batch=1` internally and reads `results.speed["inference"]`, whereas your `perf_counter()` wraps the full `predict()` call on a preloaded list of NumPy images, so it reflects batched throughput rather than comparable per-image latency; for an apples-to-apples check, either time `predict()` one image at a time or run `val()` with a matching batch size, and you can see those timing paths in the [benchmark reference](https://docs.ultralytics.com/reference/utils/benchmarks/) and [`BasePredictor.stream_inference()` docs](https://docs.ultralytics.com/reference/engine/predictor/).

**Comment by jojo9-cmd** [comment]
Thank you!

**Comment by glenn-jocher** [comment]
You're welcome — glad that clarified it.
